src/controller_examples/rabbitmq_controller/proof/helper_invariants/predicate.rs

```diff
@@ -0,0 +1,157 @@
+use vstd::{multiset::*, prelude::*, string::*};
+
+verus! {
+
+pub open spec fn object_of_key_has_no_finalizers_or_timestamp_and_only_has_controller_owner_ref(key: ObjectRef, rabbitmq: RabbitmqClusterView) -> StatePred<RMQCluster> {
+    |s: RMQCluster| {
+        s.resources().contains_key(key)
+        ==> s.resources()[key].metadata.deletion_timestamp.is_None()
+            && s.resources()[key].metadata.finalizers.is_None()
+            && exists |uid: Uid| #![auto]
+            s.resources()[key].metadata.owner_references == Some(seq![OwnerReferenceView {
+                block_owner_deletion: None,
+                controller: Some(true),
+                kind: RabbitmqClusterView::kind(),
+                name: rabbitmq.metadata.name.get_Some_0(),
+                uid: uid,
+            }])
+    }
+}
+
+pub open spec fn resource_create_request_msg(key: ObjectRef) -> FnSpec(RMQMessage) -> bool {
+    |msg: RMQMessage|
+        msg.dst.is_KubernetesAPI()
+        && msg.content.is_create_request()
+        && msg.content.get_create_request().namespace == key.namespace
+        && msg.content.get_create_request().obj.metadata.name.get_Some_0() == key.name
+        && msg.content.get_create_request().obj.kind == key.kind
+}
+
+pub open spec fn resource_update_request_msg(key: ObjectRef) -> FnSpec(RMQMessage) -> bool {
+    |msg: RMQMessage|
+        msg.dst.is_KubernetesAPI()
+        && msg.content.is_update_request()
+        && msg.content.get_update_request().key() == key
+}
+
+pub open spec fn every_resource_object_in_create_request_does_the_make_method(sub_resource: SubResource, rabbitmq: RabbitmqClusterView) -> StatePred<RMQCluster>
+    recommends
+        rabbitmq.well_formed(),
+{
+    |s: RMQCluster| {
+        let key = rabbitmq.object_ref();
+        let resource_key = get_request(sub_resource, rabbitmq).key;
+        let made_object = make(sub_resource, rabbitmq, s.ongoing_reconciles()[key].local_state);
+        forall |msg: RMQMessage| {
+            &&& #[trigger] s.network_state.in_flight.contains(msg)
+            &&& resource_create_request_msg(resource_key)(msg)
+        } ==> {
+            &&& at_rabbitmq_step(key, RabbitmqReconcileStep::AfterKRequestStep(ActionKind::Create, sub_resource))(s)
+            &&& RMQCluster::pending_k8s_api_req_msg(s, key)
+            &&& msg == s.ongoing_reconciles()[key].pending_req_msg.get_Some_0()
+            &&& made_object.is_Ok()
+            &&& msg.content.get_create_request().obj == made_object.get_Ok_0()
+        }
+            // A reminder: The last predicate implies:
+            // && msg.content.get_create_request().obj.metadata.owner_references == Some(seq![rabbitmq.controller_owner_ref()])
+    }
+}
+
+/// We have to impose constraint on the resource version of update request message now instead of making no distinction
+/// to update request messages before. This is because we don't update the whole spec/data for the object now (we only update
```

Comment:
> we don't update the whole spec/data for the object now

My understanding is that for the other fields (which are supposed to be constant, like selector of stateful set) we are going to prove a "safety" invariant saying that as long as the objects exist these fields equal X. Are you trying to do the same reasoning here (I didn't fully understand your comment here).